Normalize paths before pattern matching on OS X

edgewood opened the pull request:
The OS X file system HFS+ stores path names as Unicode, and converts
them to a variant of Unicode NFD for storage.  Because path names will
always be in this canonical form, it's not friendly to require users to
match this form exactly.  Convert paths from the repository and patterns
from the command line to NFD before comparing them.

Unix (and Windows, I think) file systems don't convert path names into a
canonical form, so users will continue to have to exactly match the path
name they want, because there could be two paths with the same character
visually that are actually composed of different byte sequences.

*Changes since last PR*
- rebased on upstream master
- fixed test error
- removed 'normalize' parameter in favor of sys.platform testing
- reorganized tests into OSX/non-OSX tests (OSX tests are temporarily using monkey patching to be able to run on non-OSX, since I don't have an OSX system for testing)
- made matching updates to ExcludePattern, including adding tests

I'm still not completely happy with the code.  The (almost) duplication in the body of the two init methods bothers me.  I could change the implementation to move the match method back to being a declared class-level method (instead of a constructed method created in init) with a decorator that calls a normalization function.  That would cost an extra function call per .match invocation, even on non-OSX, but might be worthwhile for maintainability.

ThomasWaldmann commented on borg/testsuite/helpers.py:
```diff
@@ -178,6 +179,99 @@ def test(self):
+        assert not e.match("ba/foo")
+        assert e.match(str(b"ba\x80/foo", 'latin1'))
+
+#@pytest.mark.skipif(sys.platform.startswith('darwin'), reason='OS X only test')
```
skipif ... darwin -> should be rather skipif ... not darwin.

ThomasWaldmann commented on borg/testsuite/helpers.py:
```diff
@@ -178,6 +179,99 @@ def test(self):
+        assert not e.match("ba/foo")
+        assert e.match(str(b"ba\x80/foo", 'latin1'))
+
+
```
maybe the non-OSX tests could be also doing include and excluse in one method, as seen in the above 20 lines? So there is a more direct correspondance between the test code for both platforms.

ThomasWaldmann commented on borg/helpers.py:
```diff
@@ -227,10 +233,22 @@ class IncludePattern:
+            # repository paths will be mostly in NFD, as the OSX exception list
+            # to NFD is small, so normalize to that form for best performance
+            pattern = unicodedata.normalize("NFD", pattern)
+            self.match = lambda p: match(unicodedata.normalize("NFD", p))
```
maybe there could be just 2 match functions get defined above: match and match_normalized and here, we would just normalize the pattern and assign self.match = match_normalized.

ThomasWaldmann commented on borg/helpers.py:
```diff
@@ -241,17 +259,30 @@ class ExcludePattern(IncludePattern):
+            # repository paths will be mostly in NFD, as the OSX exception list
+            # to NFD is small, so normalize to that form for best performance
+            self.pattern = unicodedata.normalize("NFD", self.pattern)
+            self.match = lambda p: match(unicodedata.normalize("NFD", p))
```
same here

ThomasWaldmann commented:
someone please testing this on OS X (adapt the hack in the code, see comments)?

edgewood commented:
I tried the match and match_normalized method, but it led to the same amount of duplicate code.  I figured out a way to do the decorator that wouldn't lead to a runtime cost for the non-OSX path, and it looks pretty good from a maintainability standpoint, too.

I implemented the suggestions in the other comments.

I feel good enough about the PR that I dropped the "don't merge this!" comment.
